protocols/request-response/src/lib.rs

```diff
@@ -396,6 +396,33 @@ where
         request_id
     }
 
+    /// TODO: Document
+    pub fn send_request_with_new_port(
```

Comment:
This doesn't really make any sense. `libp2p-request-response` hides the idea of connections and streams from the user. At the moment, there is no guarantee that a certain messages is actually going to be sent over a specific connection.

I think to effectively utilize this API in AutoNAT, we will have to rewrite it to not use `libp2p-request-response`.

The flow would something like:

- Issue a new `ToSwarm::Dial` that uses the `new_port` option
- Grab the (future) `ConnectionId` from `DialOpts`
- Remember it in the behaviour, together with the request that we want to send on _that_ connection
- Wait for `handle_established_outbound_connection` with this new `ConnectionId`
- Initialize the handler with that connection

This is quite specific to AutoNAT and I am not sure we should extend `libp2p-request-response` to be able to do this. We _could_ make `libp2p-request-response` do such kind of things as well. But, I think it is a bad fit for the abstraction because it gets too leaky. To me, `libp2p-request-response` is all about providing convenience when you just want to declare a message type and send it back and forth between two peers. If we don't hide certain details about the internals of `libp2p` with it, then the _complexity_ for the user doesn't actually go away.

Another thing that bothers me is that we expose functions such as `add_address`. Those should go away in my opinion.

Reply:
> This is quite specific to AutoNAT and I am not sure we should extend libp2p-request-response to be able to do this.

Agreed.

Reply:
Note that even with rewriting it, the issue of "don't run other protocols over that connection" is not solved although I am not entirely convinced it is actually a problem.